pepper-apis/studybuilder-cli/tenants/register-user-in-pepper-action.js

```diff
@@ -0,0 +1,368 @@
+      pepper_params.mode = event.request.body.mode;
+      console.log('Registration Mode passed in (via body) = ' + pepper_params.mode);
+    } else {
+      console.log('No Registration Mode passed in request');
+    }
+
+    if (pepper_params.mode && pepper_params.mode === "login" &&
+      !event.user.app_metadata.user_guid && !event.user.app_metadata.pepper_user_guids &&
+      !pepper_params.tempUserGuid
+    ) {
+      console.log('User not registered');
+      const loginErrPayload = {
+        code: 'user_not_registered',
+        message: 'User need to register first in order to login',
+        statusCode: 403,
+      };
+      return api.access.deny(loginErrPayload.message);
+    }
+
+    if (event.request.query.invitation_id) {
+      pepper_params.invitationId = event.request.query.invitation_id;
+      console.log('Invitation id passed in (via query) = ' + pepper_params.invitationId);
+    } else if (event.request.body.invitation_id) {
+      pepper_params.invitationId = event.request.body.invitation_id;
+      console.log('Invitation id passed in (via body) = ' + pepper_params.invitationId);
+    }
+
+    if (event.request.query.language) {
+      pepper_params.languageCode = event.request.query.language;
+      console.log('User language passed in (via query) = ' + pepper_params.languageCode);
+    } else if (event.request.body.language) {
+      pepper_params.languageCode = event.request.body.language;
+      console.log('User language passed in (via body) = ' + pepper_params.languageCode);
+    }
+
+    if (event.request.query.time_zone) {
+      pepper_params.timeZone = event.request.query.time_zone;
+      console.log('User timezone passed in (via query) = ' + pepper_params.timeZone);
+    } else if (event.request.body.time_zone) {
+      pepper_params.timeZone = event.request.body.time_zone;
+      console.log('User timezone passed in (via body) = ' + pepper_params.timeZone);
+    }
+
+    // This is the token renewal case. Let's avoid going through pepper registration
+    if (event.request.query.renew_token_only) {
+      console.log('Action request.query.renew_token_only: ');
+      api.idToken.setCustomClaim(pepperUserGuidClaim, event.user.app_metadata.user_guid);
+      return;
+    }
+
+    // In order to get a refresh token, the client must go through one of the other methods
+    // to get an id/access token first (oauth2-password, oidc-implicit-profile, etc).
+    // This allows us to assume that, if they have a refresh token, they must have gone through the
+    // registration process.
+    //
+    // Errata: This assumption is not valid in a multi-study world. It's possible to fetch a token
+    //  for one study initially, then attempt to renew it when accessing another. For the moment,
+    //  assume that, if a study guid is included with the call, the client wants us to call the
+    //  registration endpoint.
+    //var isRefreshTokenExchange = event.transaction.protocol  === "oauth2-refresh-token"; //todo
```

Comment:
todo: need to identify if refreshTokenExchange and enable the check.
Doesn't seem to impact registration and logins though.